### PR TITLE
Add matrix free Navier Stokes operator

### PR DESCRIPTION
**Pull request by lpsaavedra**
# Description of the problem

This PR is a follow up on the matrix free Navier-Stokes solver shell created in PR #755. It adds a first working version of the matrix free Navier-Stokes solver in Lethe by implementing a matrix free operator. 

# Description of the solution
- The application was renamed and updated according to the new dimension parameter added in Lethe recently.
- For the matrix free operators, a base class was added and a SUPG-PSPG derived class was implemented.
- The MF Navier-Stokes solver was updated to use the system operator instead of the system matrix as in the traditional Navier Stokes solver. 

# How Has This Been Tested?

The application is tested in 2D and 3D by using a MMS. Two element orders are considered in 2D and one in 3D. The following tests were added:

- [x] applications_tests/mf_navier_stokes/mms2d_fe1.prm
- [x] applications_tests/mf_navier_stokes/mms2d_fe2.prm
- [x] applications_tests/mf_navier_stokes/mms3d_fe1.prm

# Documentation

- No documentation has been added as the solver is not fully functional yet from the user point of view. 

# Future changes

- At the moment, the solver supports only GMRES without preconditioning as linear solver. The next step is to implement the different MG preconditioners. This will lead to the refactoring of linear solvers throughout Lethe.

### CODE REVIEW

File: include/solvers/mf_navier_stokes.h [hunk 1]
```diff
@@ -146,30 +150,22 @@ class MFGLSNavierStokesSolver
...
-  void
-  setup_GMG();
-
-  /**
-   * Members
-   */
 protected:
-  // TODO create operator: NSOperator<dim, double> system_matrix;
+  // Matrix-free operator
+  NavierStokesOperatorBase<dim, double> *system_operator;
```
Comment (peterrum) [inline include/solvers/mf_navier_stokes.h:168]: Raw pointer?

File: include/solvers/mf_navier_stokes_operators.h [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+   * classes.
+   *
+   * @param integrator FEEvaluation object that allows to evaluate functions at
+   * quadrature points and perform cell integrations.
+   */
+  virtual void
+  do_cell_integral_local(FECellIntegrator &integrator) const;
+
+  /**
+   * @brief Loop over all cell batches withing certain range and perform a cell integral with access to global vectors, i.e., gathering and scattering values.
```
Comment (peterrum) [inline include/solvers/mf_navier_stokes_operators.h:242]: Wrap line.

File: source/solvers/navier_stokes_base.cc [hunk 3]
```diff
@@ -1456,6 +1456,9 @@ NavierStokesBase<dim, VectorType, DofsType>::postprocess_fd(bool firstIter)
           // Update the time of the exact solution to the actual time
           this->exact_solution->set_time(
             simulation_control->get_current_time());
+
+          present_solution.update_ghost_values();
```
Comment (peterrum) [inline source/solvers/navier_stokes_base.cc:1460]: Could you `zero_out_ghost_values()` at the end?

File: source/solvers/mf_navier_stokes.cc [hunk 4]
```diff
@@ -408,46 +442,111 @@ MFGLSNavierStokesSolver<dim>::define_zero_constraints()
...
+  const double absolute_residual =
+    this->simulation_parameters.linear_solver.minimum_residual;
+  const double relative_residual =
+    this->simulation_parameters.linear_solver.relative_residual;
+
+  if (this->simulation_parameters.linear_solver.solver ==
+      Parameters::LinearSolver::SolverType::gmres)
+    solve_system_GMRES(initial_step, absolute_residual, relative_residual);
+  else
+    throw(std::runtime_error("This solver is not allowed"));
```
Comment (peterrum) [inline source/solvers/mf_navier_stokes.cc:457]: AsserThrow?

File: source/solvers/mf_navier_stokes_operators.cc [hunk 5]
```diff
@@ -0,0 +1,810 @@
...
+          const unsigned int count_global =
+            Utilities::MPI::sum(count, dof_handler.get_communicator());
+
+          const unsigned int count_cells_global =
+            Utilities::MPI::sum(matrix_free.n_cell_batches(),
+                                dof_handler.get_communicator());
+
+          if (Utilities::MPI::this_mpi_process(
+                dof_handler.get_communicator()) == 0)
+            std::cout << count_global << " " << count_cells_global << std::endl;
```
Comment (peterrum) [inline source/solvers/mf_navier_stokes_operators.cc:194]: Shouldn't this be optional?

File: source/solvers/mf_navier_stokes_operators.cc [hunk 6]
```diff
@@ -0,0 +1,810 @@
...
+    }
+}
+
+template <int dim, typename number>
+void
+NavierStokesOperatorBase<dim, number>::compute_element_size()
+{
+  FECellIntegrator phi(matrix_free);
+
+  unsigned int n_cells =
```
Comment (peterrum) [inline source/solvers/mf_navier_stokes_operators.cc:205]:
```suggestion
  const unsigned int n_cells =
```

**Review by peterrum — APPROVED** [review]

File: include/solvers/mf_navier_stokes.h (hunk 1 above)
Comment (blaisb) [reply include/solvers/mf_navier_stokes.h:168]: std::shared would be cleaner here

File: include/solvers/mf_navier_stokes_operators.h [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+protected:
+  MatrixFree<dim, number>                matrix_free;
+  AffineConstraints<number>              constraints;
+  mutable TrilinosWrappers::SparseMatrix system_matrix;
+  AlignedVector<VectorizedArray<number>> element_size;
+  unsigned int                           fe_degree;
+  const Function<dim> *                  forcing_function;
+  double                                 viscosity;
+
+  // Variables needed from the last newton step vector
```
Comment (blaisb) [inline include/solvers/mf_navier_stokes_operators.h:295]:
```suggestion
  // Variables needed from the last Newton step vector
```

File: include/solvers/mf_navier_stokes_operators.h [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+ * @tparam dim An integer that denotes the number of spatial dimensions.
+ * @tparam number Abstract type for number across the class (i.e., double).
+ */
+template <int dim, typename number>
+class NavierStokesSUPGPSPGOperator
+  : public NavierStokesOperatorBase<dim, number>
+{
+public:
+  using FECellIntegrator = FEEvaluation<dim, -1, 0, dim + 1, number>;
+  using VectorType       = LinearAlgebra::distributed::Vector<number>;
```
Comment (blaisb) [inline include/solvers/mf_navier_stokes_operators.h:326]: Does the using need to be remade for every inherited class?

File: source/solvers/mf_navier_stokes.cc [hunk 9]
```diff
@@ -199,49 +217,65 @@ MFGLSNavierStokesSolver<dim>::update_boundary_conditions()
...
+        "Type of initial condition is not supported by MF Navier-Stokes");
+    }
 }
 
 template <int dim>
 void
-MFGLSNavierStokesSolver<dim>::assemble_system_matrix()
+MFNavierStokesSolver<dim>::assemble_system_matrix()
 {
   // TODO
```
Comment (blaisb) [inline source/solvers/mf_navier_stokes.cc:249]: Is there really a TODO here? it should be a blank function forever no?

File: source/solvers/mf_navier_stokes.cc (hunk 4 above)
Comment (blaisb) [reply source/solvers/mf_navier_stokes.cc:457]: yup should be

File: source/solvers/mf_navier_stokes_operators.cc (hunk 5 above)
Comment (blaisb) [reply source/solvers/mf_navier_stokes_operators.cc:194]: At least if the output is not optional, it should be a bit more verbose in what it means

File: source/solvers/mf_navier_stokes_operators.cc [hunk 10]
```diff
@@ -0,0 +1,810 @@
...
+    matrix_free.n_cell_batches() + matrix_free.n_ghost_cell_batches();
+  element_size.resize(n_cells);
+
+  for (unsigned int cell = 0; cell < n_cells; ++cell)
+    {
+      for (auto lane = 0u;
+           lane < matrix_free.n_active_entries_per_cell_batch(cell);
+           lane++)
+        {
+          double h_k = matrix_free.get_cell_iterator(cell, lane)->measure();
```
Comment (blaisb) [inline source/solvers/mf_navier_stokes_operators.cc:215]:
```suggestion
          const double h_k = matrix_free.get_cell_iterator(cell, lane)->measure();
```

File: source/solvers/mf_navier_stokes_operators.cc [hunk 11]
```diff
@@ -0,0 +1,810 @@
...
+      auto previous_hessian_diagonal =
+        this->nonlinear_previous_hessian_diagonal(cell, q);
+
+      // Calculate tau
+      VectorizedArray<number> u_mag = VectorizedArray<number>(0.0);
+      VectorizedArray<number> tau   = VectorizedArray<number>(0.0);
+      for (unsigned int k = 0; k < dim; ++k)
+        {
+          u_mag += previous_values[k];
+        }
```
Comment (blaisb) [inline source/solvers/mf_navier_stokes_operators.cc:593]:
```suggestion
      for (unsigned int k = 0; k < dim; ++k)
          u_mag += previous_values[k];
```
one line for loop are cleaning I think in this case

**Review by blaisb — CHANGES_REQUESTED** [review]
Tiny change. Once addressed this will be easy to merge

File: source/solvers/mf_navier_stokes_operators.cc (hunk 5 above)
Comment (lpsaavedra) [reply source/solvers/mf_navier_stokes_operators.cc:194]: Yes, it should be optional and I will add a comment explaining the purpose of it.

File: include/solvers/mf_navier_stokes_operators.h (hunk 8 above)
Comment (lpsaavedra) [reply include/solvers/mf_navier_stokes_operators.h:326]: I think they cannot be inherited as it is right now. The only way would be to declare both classes inside a namespace and declare the using directive right before it.

**Review by blaisb — APPROVED** [review]

**Review by peterrum — APPROVED** [review]

**Pull request merged**